### ISSUE TITLE
Security or ease?

### ISSUE THREAD

**Issue by krainboltgreene** [discourse/onebox#166]
tl;dr: Security issues are dangerous and we can either allow `&` in our links and not care that it gets turned into `&amp;` or we unescape urls and write code to prevent XSS problems. Most of this is lead up, last bit is the current issue.

We have a unified entry point for all URLs being processed by onebox:

``` ruby
Onebox.preview(url)
```

Normally any client that passes data to `Onebox.preview()` should be able to cull out non-url strings, but for the sake of security auditing lets assume you were able to get _any_ string to this point:

``` ruby
string = "anything"
Onebox.preview(string)
```

This example is harmless, since it'd not match any `Onebox::Engine` classes, but lets look at this example:

``` ruby
string = "http://amazon.com http://badguy.com"
Onebox.preview(string)
```

This example will get through our system into the template and will actually match the `Onebox::Engine::AmazonOnebox`. For reference our matcher for that onebox is: `/^http(?:s)?:\/\/(?:www\.)?(?:(?:\w)+\.)?(amazon)\.(?:com|ca)(?:\/)?/`.

Now we've fixed the Amazon one (just recently) but I wanted to use this example to show easy it is to let an XSS problem slip through.

So now lets talk about injecting html into the string:

``` ruby
string = "http://amazon.com/product/1\" onclick=\"alert('BOO!')\""
Onebox.preview(string)
```

Luckily the `open-uri` stdlib catches this and throws `URI::InvalidURIError: bad URI(is not URI?): http://amazon.com/product/1" onclick="alert('BOO!')"`. If it didn't you'd still be saved here:

``` ruby
Mustache.render("{{url}}", url: string)
"http://amazon.com/product/1&quot; onclick=&quot;alert(&#39;BOO!&#39;)&quot;"
```

Thanks to the default escaping of things in mustache's interpolation. This however causes a few problems we didn't expect. Consider this URL:

``` ruby
string = "http://www.linkedin.com/profile/view?id=162125271&trk=nav_responsive_tab_profile"
Mustache.render("{{url}}", url: string)
"http://www.linkedin.com/profile/view?id=162125271&amp;trk=nav_responsive_tab_profile"
```

This URL is no longer the same as the one given to us and the only way to avoid this is to unescape the url given to the template, but that opens the floodgates to other abuses.

/cc @dysania/owners @eviltrout @SamSaffron
